File: src/monte_carlo.c
#include "monte_carlo.h"

// Box-Muller transform to generate Gaussian random numbers
float generateGaussianRandom(void) {
    static int has_spare = 0;
    static float spare;
    
    if (has_spare) {
        has_spare = 0;
        return spare;
    }
    
    has_spare = 1;
    float u = 2.0 * rand() / RAND_MAX - 1.0;
    float v = 2.0 * rand() / RAND_MAX - 1.0;
    float s = u * u + v * v;
    
    while (s >= 1.0 || s == 0.0) {
        u = 2.0 * rand() / RAND_MAX - 1.0;
        v = 2.0 * rand() / RAND_MAX - 1.0;
        s = u * u + v * v;
    }
    
    float multiplier = sqrt(-2.0 * log(s) / s);
    spare = v * multiplier;
    return u * multiplier;
}

// Main Monte Carlo simulation using Geometric Brownian Motion
void runMonteCarloSimulation(SimulationParams params, SimulationResults *results) {
    float dt = 1.0 / 252.0;  // Daily time step (252 trading days/year)
    float drift = params.expected_return * dt;
    float diffusion = params.volatility * sqrt(dt);
    
    printf("\n🚀 Starting Monte Carlo Simulation...\n");
    printf("   Initial Price: $%.2f\n", params.initial_price);
    printf("   Expected Return: %.2f%%\n", params.expected_return * 100);
    printf("   Volatility: %.2f%%\n", params.volatility * 100);
    printf("   Time Steps: %d days\n", params.time_steps);
    printf("   Simulations: %d\n\n", params.num_simulations);
    
    // Run Monte Carlo simulations
    for (int sim = 0; sim < params.num_simulations; sim++) {
        float price = params.initial_price;
        
        // Simulate daily price changes using Geometric Brownian Motion
        // dS = mu * S * dt + sigma * S * dW
        for (int t = 0; t < params.time_steps; t++) {
            float random_shock = generateGaussianRandom();
            float log_return = drift + diffusion * random_shock;
            price = price * exp(log_return);
        }
        
        results->final_prices[sim] = price;
        
        // Progress indicator
        if ((sim + 1) % 1000 == 0) {
            printf("   ✓ Completed %d simulations\n", sim + 1);
        }
    }
    
    printf("\n✅ Simulation complete!\n");
}

// Calculate statistical metrics from results
void calculateStatistics(SimulationResults *results) {
    // Find min and max
    results->min_price = results->final_prices[0];
    results->max_price = results->final_prices[0];
    results->mean_price = 0;
    
    for (int i = 0; i < MAX_SIMULATIONS; i++) {
        if (results->final_prices[i] < results->min_price) {
            results->min_price = results->final_prices[i];
        }
        if (results->final_prices[i] > results->max_price) {
            results->max_price = results->final_prices[i];
        }
        results->mean_price += results->final_prices[i];
    }
    
    results->mean_price /= MAX_SIMULATIONS;
    
    // Calculate standard deviation
    float variance = 0;
    for (int i = 0; i < MAX_SIMULATIONS; i++) {
        float diff = results->final_prices[i] - results->mean_price;
        variance += diff * diff;
    }
    variance /= MAX_SIMULATIONS;
    results->std_dev = sqrt(variance);
    
    // Sort prices for percentile calculations
    float sorted[MAX_SIMULATIONS];
    for (int i = 0; i < MAX_SIMULATIONS; i++) {
        sorted[i] = results->final_prices[i];
    }
    
    // Simple bubble sort
    for (int i = 0; i < MAX_SIMULATIONS - 1; i++) {
        for (int j = 0; j < MAX_SIMULATIONS - i - 1; j++) {
            if (sorted[j] > sorted[j + 1]) {
                float temp = sorted[j];
                sorted[j] = sorted[j + 1];
                sorted[j + 1] = temp;
            }
        }
    }
    
    // Calculate confidence intervals
    int idx_5 = (int)(MAX_SIMULATIONS * 0.05);
    int idx_95 = (int)(MAX_SIMULATIONS * 0.95);
    int idx_1 = (int)(MAX_SIMULATIONS * 0.01);
    int idx_99 = (int)(MAX_SIMULATIONS * 0.99);
    
    results->confidence_95_lower = sorted[idx_5];
    results->confidence_95_upper = sorted[idx_95];
    results->confidence_99_lower = sorted[idx_1];
    results->confidence_99_upper = sorted[idx_99];
    
    // Calculate Value at Risk (95%)
    results->value_at_risk = results->mean_price - results->confidence_95_lower;
    
    // Calculate probability of profit/loss
    int profit_count = 0;
    for (int i = 0; i < MAX_SIMULATIONS; i++) {
        if (results->final_prices[i] > results->final_prices[0]) {
            profit_count++;
        }
    }
    results->prob_profit = (float)profit_count / MAX_SIMULATIONS * 100;
    results->prob_loss = 100 - results->prob_profit;
}

// Print results to console
void printResults(SimulationResults results, SimulationParams params) {
    printf("\n");
    printf("╔════════════════════════════════════════════════════════════╗\n");
    printf("║         MONTE CARLO SIMULATION RESULTS                    ║\n");
    printf("╚════════════════════════════════════════════════════════════╝\n\n");
    
    printf("📊 PRICE STATISTICS:\n");
    printf("   Initial Price:          $%.2f\n", params.initial_price);
    printf("   Mean Final Price:       $%.2f\n", results.mean_price);
    printf("   Minimum Price:          $%.2f\n", results.min_price);
    printf("   Maximum Price:          $%.2f\n", results.max_price);
    printf("   Standard Deviation:     $%.2f\n\n", results.std_dev);
    
    printf("📈 CONFIDENCE INTERVALS:\n");
    printf("   95%% CI: [$%.2f - $%.2f]\n", results.confidence_95_lower, results.confidence_95_upper);
    printf("   99%% CI: [$%.2f - $%.2f]\n\n", results.confidence_99_lower, results.confidence_99_upper);
    
    printf("💰 PROFIT/LOSS ANALYSIS:\n");
    printf("   Probability of Profit:  %.2f%%\n", results.prob_profit);
    printf("   Probability of Loss:    %.2f%%\n", results.prob_loss);
    printf("   Value at Risk (95%%):    $%.2f\n\n", results.value_at_risk);
    
    printf("📉 RETURNS:\n");
    float expected_return = ((results.mean_price - params.initial_price) / params.initial_price) * 100;
    float best_return = ((results.max_price - params.initial_price) / params.initial_price) * 100;
    float worst_return = ((results.min_price - params.initial_price) / params.initial_price) * 100;
    
    printf("   Expected Return:       %.2f%%\n", expected_return);
    printf("   Best Case Return:      %.2f%%\n", best_return);
    printf("   Worst Case Return:     %.2f%%\n\n", worst_return);
}

// Save results to CSV file
void saveResultsToFile(SimulationResults results, SimulationParams params, const char *filename) {
    FILE *file = fopen(filename, "w");
    
    if (file == NULL) {
        printf("Error opening file %s\n", filename);
        return;
    }
    
    fprintf(file, "Monte Carlo Simulation Results\n");
    fprintf(file, "Initial Price,%.2f\n", params.initial_price);
    fprintf(file, "Expected Return,%.2f%%\n", params.expected_return * 100);
    fprintf(file, "Volatility,%.2f%%\n\n", params.volatility * 100);
    
    fprintf(file, "Statistical Metrics\n");
    fprintf(file, "Mean Final Price,%.2f\n", results.mean_price);
    fprintf(file, "Min Price,%.2f\n", results.min_price);
    fprintf(file, "Max Price,%.2f\n", results.max_price);
    fprintf(file, "Standard Deviation,%.2f\n", results.std_dev);
    fprintf(file, "Value at Risk (95%%),%.2f\n", results.value_at_risk);
    fprintf(file, "Probability of Profit,%.2f%%\n\n", results.prob_profit);
    
    fprintf(file, "Final Prices\n");
    for (int i = 0; i < MAX_SIMULATIONS; i++) {
        fprintf(file, "%.2f\n", results.final_prices[i]);
    }
    
    fclose(file);
    printf("✓ Results saved to %s\n", filename);
}

File: frontend/src/lib/monteCarlo.ts
export interface SimulationParams {
  initialPrice: number;
  expectedReturn: number;
  volatility: number;
  timeSteps: number;
  numSimulations: number;
}

export interface SimulationResults {
  meanPrice: number;
  minPrice: number;
  maxPrice: number;
  stdDev: number;
  confidence95Lower: number;
  confidence95Upper: number;
  confidence99Lower: number;
  confidence99Upper: number;
  valueAtRisk: number;
  probProfit: number;
  probLoss: number;
  // We'll keep a subset of full paths for charting
  chartPaths: number[][]; 
}

// Box-Muller transform to generate Gaussian random numbers
function generateGaussianRandom(): number {
  let u = 0, v = 0;
  while(u === 0) u = Math.random(); // Converting [0,1) to (0,1)
  while(v === 0) v = Math.random();
  return Math.sqrt(-2.0 * Math.log(u)) * Math.cos(2.0 * Math.PI * v);
}

export function runMonteCarloSimulation(params: SimulationParams): SimulationResults {
  const dt = 1.0 / 252.0; // Daily time step (252 trading days/year)
  const drift = params.expectedReturn * dt;
  const diffusion = params.volatility * Math.sqrt(dt);
  
  const finalPrices: number[] = new Float32Array(params.numSimulations) as any;
  const chartPaths: number[][] = [];
  const MAX_CHART_PATHS = 50; // Only keep 50 paths for the UI chart to stay performant

  let meanPrice = 0;
  let minPrice = Infinity;
  let maxPrice = -Infinity;
  let profitCount = 0;

  for (let sim = 0; sim < params.numSimulations; sim++) {
    let price = params.initialPrice;
    const currentPath: number[] = [price];
    
    // Simulate daily price changes using Geometric Brownian Motion
    for (let t = 0; t < params.timeSteps; t++) {
      const randomShock = generateGaussianRandom();
      const logReturn = drift + diffusion * randomShock;
      price = price * Math.exp(logReturn);
      if (sim < MAX_CHART_PATHS) {
        currentPath.push(price);
      }
    }
    
    finalPrices[sim] = price;
    
    if (sim < MAX_CHART_PATHS) {
      chartPaths.push(currentPath);
    }

    meanPrice += price;
    if (price < minPrice) minPrice = price;
    if (price > maxPrice) maxPrice = price;
    if (price > params.initialPrice) profitCount++;
  }

  meanPrice /= params.numSimulations;

  // Calculate Standard Deviation
  let variance = 0;
  for (let i = 0; i < params.numSimulations; i++) {
    const diff = finalPrices[i] - meanPrice;
    variance += diff * diff;
  }
  variance /= params.numSimulations;
  const stdDev = Math.sqrt(variance);

  // Sort prices for percentiles
  const sortedPrices = Float32Array.from(finalPrices as any);
  sortedPrices.sort();

  const idx5 = Math.floor(params.numSimulations * 0.05);
  const idx95 = Math.floor(params.numSimulations * 0.95);
  const idx1 = Math.floor(params.numSimulations * 0.01);
  const idx99 = Math.floor(params.numSimulations * 0.99);

  const confidence95Lower = sortedPrices[idx5];
  const confidence95Upper = sortedPrices[idx95];
  const confidence99Lower = sortedPrices[idx1];
  const confidence99Upper = sortedPrices[idx99];

  const valueAtRisk = meanPrice - confidence95Lower;
  const probProfit = (profitCount / params.numSimulations) * 100;
  const probLoss = 100 - probProfit;

  return {
    meanPrice,
    minPrice,
    maxPrice,
    stdDev,
    confidence95Lower,
    confidence95Upper,
    confidence99Lower,
    confidence99Upper,
    valueAtRisk,
    probProfit,
    probLoss,
    chartPaths,
  };
}
